Fixed #24553 -- Added the list of applications to the context of admin views

xrmx opened the pull request:
Pass the list of available applications to more views so it is possible to show a menu with them.

Discussion on list:
​https://groups.google.com/forum/#!searchin/django-developers/having$20the$20app/django-developers/SLwh7zJmkTQ/RPUABdNYuCIJ

timgraham commented on tests/timezones/urls.py:
```diff
@@ -1,8 +1,9 @@
 from django.conf.urls import include, url
-from django.contrib import admin
+
+from .admin import site
```
this line goes under `from . import admin` with no blankline between.

xrmx commented:
Changes since last pull requesst:
- fixed broken english
- added missing test for index
- for the change form view moved adding available_apps fromfrom render_change_form
- s/build_app_list/get_app_list as Tim suggested in the ticket

timgraham commented:
I merged the first two commits. The views that `available_apps` is added to still seems a bit arbitrary. Did you consider adding it to the default `AdminSite.each_context()` method?

xrmx commented:
Fine for me, updated patch. Thanks!

timgraham commented:
Doc updates are also required: https://docs.djangoproject.com/en/dev/ref/contrib/admin/#django.contrib.admin.AdminSite.each_context
Maybe we can test `AdminSite.each_context()` directly, instead of the current tests.

xrmx commented:
Am trying to update the docs but this
+    * ``available_apps``: a list of applications from the
+        :ref:`application registry</ref/applications/>` available for the current
+        session

gives this warning
/home/rm/src/django/docs/ref/contrib/admin/index.txt:2509: WARNING: undefined label: /ref/applications/ (if the link has no caption the label must precede a section header)

Any hint?

xrmx commented:
Pushed with updated tests, still documentation build does not look correct.

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2505,10 +2505,13 @@ Templates can override or extend base admin templates as described in
     * ``site_url``: :attr:`AdminSite.site_url`
     * ``has_permission``: :meth:`AdminSite.has_permission`
+    * ``available_apps``: a list of applications from the
+        :ref:`application registry</ref/applications/>` available for the current
```
should be doc instead of ref
also a line break is allowed inside that reference between words so you can make the previous line a bit longer

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2505,10 +2505,13 @@ Templates can override or extend base admin templates as described in
+        :ref:`application registry</ref/applications/>` available for the current
+        session
 
-    .. versionchanged:: 1.8
```
please don't delete the note for 1.8 (we remove them after 2 releases)

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -6166,3 +6167,34 @@ def test_implicitly_generated_pk(self):
+        request = self.factory.get(reverse('admin:index'))
+        request.user = self.u1
+        ctx = site.each_context(request)
+        self.assertIn('site_header', ctx)
```
Might not be a bad idea to assert the value of each key.

xrmx replied:
Am sorry but i don't understand this comment, could you please elaborate a bit?

timgraham replied:
We can skip the "assertIn" assertions I think.

timgraham commented:
Please add to release notes as well.

xrmx commented:
Pushed a new version with the docs fixed, release notes updated, tested values of keys, for the available_apps i've just checked that it's not empty.

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2505,11 +2505,18 @@ Templates can override or extend base admin templates as described in
     * ``site_url``: :attr:`AdminSite.site_url`
     * ``has_permission``: :meth:`AdminSite.has_permission`
+    * ``available_apps``: a list of applications from the
+        :doc:`application registry </ref/applications/>` available for the
```
deindent two spaces

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2505,11 +2505,18 @@ Templates can override or extend base admin templates as described in
     * ``has_permission``: :meth:`AdminSite.has_permission`
+    * ``available_apps``: a list of applications from the
+        :doc:`application registry </ref/applications/>` available for the
+        current session
```
add period?

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -6166,3 +6167,39 @@ def test_implicitly_generated_pk(self):
+        self.assertIn('has_permission', ctx)
+        self.assertEqual(ctx['has_permission'], True)
+        self.assertIn('available_apps', ctx)
+        self.assertNotEqual(ctx['available_apps'], [])
```
If we are documenting this, I think we should try to add assertions for its contents to prevent breaking it. Also I think the documentation should be more specific about what it contains.

xrmx commented:
Updated patch:
- fixed doc indentation, added period
- added test to check actual content of the available_apps list

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2507,11 +2507,18 @@ Templates can override or extend base admin templates as described in
     * ``site_title``: :attr:`AdminSite.site_title`
     * ``site_url``: :attr:`AdminSite.site_url`
     * ``has_permission``: :meth:`AdminSite.has_permission`
+    * ``available_apps``: a list of applications from the
```
"a list of applications" is too vague, I think. Looking at the test, it seems to return much more than a list of simple items. Maybe an example of the structure that is returned would be appropriate?

timgraham commented on docs/releases/1.9.txt:
```diff
@@ -48,6 +48,11 @@ Minor features
+* A new context variable ``available_apps``, which lists the
+  :doc:`application registry </ref/applications/>` available for the current
+  session, has been added to the :meth:`AdminSite.each_context(request)
+  <django.contrib.admin.AdminSite.each_context>`.
```
to the AdminSite.each_context() method (no need to list arguments and add "method")

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -6166,3 +6167,70 @@ def test_implicitly_generated_pk(self):
+    PASSWORD_HASHERS=['django.contrib.auth.hashers.SHA1PasswordHasher'],
+    ROOT_URLCONF="admin_views.urls",
+)
+class SiteEachContextTest(TestCase):
```
There has been discussion about better organizing the `admin_views` tests. How about moving this to a new file along with its admin and urlpatterns (if that works)? Something like `tests/admin_views/test_adminsite.py`

timgraham commented on tests/admin_views/tests.py:
```diff
@@ -6166,3 +6167,70 @@ def test_implicitly_generated_pk(self):
+
+    def test_each_context(self):
+        ctx = self.ctx
+        self.assertIn('site_header', ctx)
```
we can remove the `assertIn` assertions I think. If we get a `KeyError` on the `assertEqual` assertions, that's just as helpful.

xrmx commented:
Updated pull request per latest review

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2507,11 +2507,32 @@ Templates can override or extend base admin templates as described in
     * ``site_url``: :attr:`AdminSite.site_url`
     * ``has_permission``: :meth:`AdminSite.has_permission`
+    * ``available_apps``: a list of applications from the
+      :doc:`application registry </ref/applications/>` available for the
```
you can probably move part of this to the previous line and break after "registry"

xrmx replied:
I had missed this, fixed and pushed again.

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2507,11 +2507,32 @@ Templates can override or extend base admin templates as described in
+      with the following keys:
+
+      * ``app_label``: the application label
+      * ``app_url``: the url of the ``application index`` in the admin
```
url -> URL
I'd omit the backticks around application index since it's not a code thing

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2507,11 +2507,32 @@ Templates can override or extend base admin templates as described in
+
+      * ``app_label``: the application label
+      * ``app_url``: the url of the ``application index`` in the admin
+      * ``has_module_perms``: a boolean indicating module permissions
```
"indicating module permissions" could be explained better

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2507,11 +2507,32 @@ Templates can override or extend base admin templates as described in
+
+      * ``object_name``: class name of the model
+      * ``name``: plural name of the model
+      * ``perms``: a ``dict`` tracking ``add``, ``change`` and ``delete`` permissions
```
comma after "change"

timgraham commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -2507,11 +2507,32 @@ Templates can override or extend base admin templates as described in
+      * ``object_name``: class name of the model
+      * ``name``: plural name of the model
+      * ``perms``: a ``dict`` tracking ``add``, ``change`` and ``delete`` permissions
+      * ``admin_url``: admin ``change list`` url for the model
```
URL
changelist (No space, no backticks)

timgraham commented on tests/admin_views/admin.py:
```diff
@@ -979,3 +979,8 @@ def get_formsets_with_inlines(self, request, obj=None):
 site7 = admin.AdminSite(name="admin7")
 site7.register(Article, ArticleAdmin2)
+
+# Used to test available_applications available in context
```
changes in this file can be removed?

timgraham commented on tests/admin_views/test_adminsite.py:
```diff
@@ -0,0 +1,83 @@
+
+from .models import Article
+
+site6 = admin.AdminSite(name="admin6")
```
site = admin.AdminSite(name="test_adminsite")

timgraham commented on tests/admin_views/test_adminsite.py:
```diff
@@ -0,0 +1,83 @@
+            id=100, password='sha1$995a3$6011485ea3834267d719b4c801409b8b1ddd0158',
+            last_login=datetime.datetime(2007, 5, 30, 13, 20, 10), is_superuser=True, username='super',
+            first_name='Super', last_name='User', email='super@example.com',
+            is_staff=True, is_active=True, date_joined=datetime.datetime(2007, 5, 30, 13, 20, 10)
```
prefer including a trailing comma in kwargs so if more items are added in the future we don't have to modify this line again

timgraham commented on tests/admin_views/test_adminsite.py:
```diff
@@ -0,0 +1,83 @@
+class SiteEachContextTest(TestCase):
+    """
+    Check each_context contains the documented variables, check that available_apps context
+    variable structure is the expected one
```
add period

timgraham commented on tests/admin_views/test_adminsite.py:
```diff
@@ -0,0 +1,83 @@
+)
+class SiteEachContextTest(TestCase):
+    """
+    Check each_context contains the documented variables, check that available_apps context
```
check that -> and that
(no comma needed since the two clauses are independent)

timgraham commented:
This is looking good. Please bump the ticket to "ready for checkin" after those updates.

timgraham commented:
merged in bd53db5eab05099ae371348529c6428e0da95c6a, thanks!

timgraham commented:
I think this patch isn't completely isolated, as the test suite [fails when run in reverse](http://djangoci.com/job/master-trusty-reverse/). Could you investigate?

xrmx commented:
I'll try to give a look at that tomorrow

xrmx commented:
So this looks like a case of SiteAdmin instance leaking between apps, cooking a patch.

xrmx commented:
Opened a new pull request fixing this:
https://github.com/django/django/pull/4659
